cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,495 @@
+    // Starting vertex ids of each partition
+    id_segments.push_back(graph_view.get_local_adj_matrix_partition_row_last(i));
+    count_offsets.push_back(counter);
+    counter += matrix_partition.get_major_size();
+  }
+  rmm::device_uvector<vertex_t> vertex_id_segments(id_segments.size(), handle.get_stream());
+  rmm::device_uvector<vertex_t> vertex_count_offsets(count_offsets.size(), handle.get_stream());
+  raft::update_device(
+    vertex_id_segments.data(), id_segments.data(), id_segments.size(), handle.get_stream());
+  raft::update_device(
+    vertex_count_offsets.data(), count_offsets.data(), count_offsets.size(), handle.get_stream());
+
+  thrust::transform(
+    handle.get_thrust_policy(),
+    active_sources.begin(),
+    active_sources.end(),
+    active_source_degrees.begin(),
+    [seg                  = vertex_id_segments.data(),
+     global_out_degrees   = global_out_degrees.data(),
+     vertex_count_offsets = vertex_count_offsets.data(),
+     count                = vertex_id_segments.size()] __device__(auto v) {
+      // Find which partition id did the source belong to
+      auto partition_id =
+        thrust::distance(seg, thrust::upper_bound(thrust::seq, seg, seg + count, v)) - 1;
+      // starting position of the segment within global_degree_offset
+      // where the information for partition (partition_id) starts
+      //  vertex_count_offsets[partition_id]
+      // The relative location of offset information for vertex id v within
+      // the segment
+      //  v - seg[partition_id]
+      auto location_in_segment = v - seg[partition_id];
+      // read location of global_degree_offset needs to take into account the
+      // partition offsets because it is a concatenation of all the offsets
+      // across all partitions
+      auto location = location_in_segment + vertex_count_offsets[partition_id];
+
+      return global_out_degrees[location];
+    });
+  return active_source_degrees;
+}
+
+/**
+ * @brief Calculate global degree information for all vertices represented by current gpu
+ *
+ * Calculate local degree and perform row wise exclusive scan over all gpus in column
+ * communicator.
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @return Tuple of two device vectors. The first one contains per source edge-count encountered
+ * by gpus in the column communicator before current gpu. The second device vector contains the
+ * global out degree for every source represented by current gpu
+ */
+template <typename GraphViewType>
+std::tuple<rmm::device_uvector<typename GraphViewType::edge_type>,
+           rmm::device_uvector<typename GraphViewType::edge_type>>
+get_global_degree_information(raft::handle_t const& handle, GraphViewType const& graph_view)
+{
```

Comment:
What's the difference between this function and compute_out_degrees (https://github.com/rapidsai/cugraph/blob/branch-22.04/cpp/include/cugraph/graph_view.hpp#L591)?

Reply:
This gives two results. The exclusive_scan of the degrees up till that gpu and the total out degrees

Reply:
Then, can't you 1) compute "total out degrees" first and 2) sum total out degrees in each GPU, and 3) run exclusive scan? This sounds much simpler than the code here.

Reply:
Umm... please disregard, I think I misunderstood your explanation.